### PR TITLE
Implement dynamic hardware palette sizing.

### PR DESCRIPTION
**Pull request by RoosterDragon**
The HardwarePalette will now grow its palette buffer and texture in power-of-2 increments. This avoids it having to allocate memory for a full 256x256 texture up front. In practice the default mods use 22 or 23 palettes so a 32x256 texture is used. This means both the buffer and texture save neatly on memory. Additionally, HardwarePalette.ApplyModifiers sees a nice speedup as it has to transfer a much smaller amount of memory from the buffer to the texture.

To facilitate this change, the MaxPalettes constant is no longer publicly exposed. Instead the PaletteReference deals with the calculation of the index and this is passed into the appropriate methods.

### CODE REVIEW

File: OpenRA.Game/Graphics/HardwarePalette.cs [hunk 1]
```diff
@@ -16,14 +16,15 @@ namespace OpenRA.Graphics
 {
 	public sealed class HardwarePalette : IDisposable
 	{
-		public const int MaxPalettes = 256;
+		const int MaxPalettes = 256;
```
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: You might as well remove this now-arbitrary limit, too.

File: OpenRA.Game/Graphics/HardwarePalette.cs [hunk 2]
```diff
@@ -16,14 +16,15 @@ namespace OpenRA.Graphics
...
 
 		public ITexture Texture { get; private set; }
+		public int Height { get { return buffer.GetLength(0); } }
 		readonly Dictionary<string, ImmutablePalette> palettes = new Dictionary<string, ImmutablePalette>();
 		readonly Dictionary<string, MutablePalette> modifiablePalettes = new Dictionary<string, MutablePalette>();
 		readonly IReadOnlyDictionary<string, MutablePalette> readOnlyModifiablePalettes;
 		readonly Dictionary<string, int> indices = new Dictionary<string, int>();
-		readonly uint[,] buffer = new uint[MaxPalettes, Palette.Size];
+		uint[,] buffer = new uint[0, Palette.Size];
```
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Would it make sense to define a sensibly-sized default (say 8 or 16) to cut down on a few resizes?  Probably not in practice...

File: OpenRA.Game/Graphics/HardwarePalette.cs [hunk 3]
```diff
@@ -60,6 +61,14 @@ public void AddPalette(string name, ImmutablePalette p, bool allowModifiers)
 			int index = palettes.Count;
 			indices.Add(name, index);
 			palettes.Add(name, p);
+
+			if (palettes.Count > buffer.GetLength(0))
+			{
+				var oldBuffer = buffer;
+				buffer = new uint[Exts.NextPowerOf2(palettes.Count), Palette.Size];
```
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: I've never been a fan of these two dimensional arrays.  Would it be worth changing to a 1D array, and then using `Array.Resize`?

Comment (RoosterDragon) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Somebody would need to update the ```Texture.SetData(uint[,])``` method and fix all callers. I'm ambivalent to such a change, but it's a job for a new PR IMO.

File: OpenRA.Game/Graphics/HardwarePalette.cs (hunk 1 above)
Comment (RoosterDragon) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Done.

File: OpenRA.Game/Graphics/HardwarePalette.cs (hunk 2 above)
Comment (RoosterDragon) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: I don't think it's worthwhile here personally, it just doesn't happen enough.

File: OpenRA.Game/Graphics/HardwarePalette.cs (hunk 3 above)
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: We have a `Texture.SetData(byte[], int, int)`, and it looks like `IPalette.CopyToArray` should already handle copying into a `byte[]` array.  It would break on big-endian systems, but the code already fails at this in a bunch of other places.

Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: But I may be missing something, so if it does take more work than the couple of lines that i'm imagining then it would be better left to its own PR.

Comment (RoosterDragon) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Done.

File: OpenRA.Game/Graphics/HardwarePalette.cs [hunk 4]
```diff
@@ -52,14 +51,21 @@ public int GetPaletteIndex(string name)
...
 				throw new InvalidOperationException("Palette {0} has already been defined".F(name));
 
 			int index = palettes.Count;
 			indices.Add(name, index);
 			palettes.Add(name, p);
+
+			if (palettes.Count > Height)
+			{
+				Height = Exts.NextPowerOf2(palettes.Count);
+				var oldBuffer = buffer;
```
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: This could be reduced to `Array.Resize(ref buffer, Height * Palette.Size * 4);` for brevity.

Comment (RoosterDragon) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Done.

**Comment by chrisforbes** [comment]
:+1:

**Comment by pchote** [comment]
Looks good. Sorry for the delay on checking this again, but things are pretty crazy for me right now.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/6237d3db2ccbc47a29fbee78eadda99153e631e5%5E...6237d3db2ccbc47a29fbee78eadda99153e631e5
